File: android/app/src/main/java/ws/baseline/autopilot/geo/LandingZone.java
package ws.baseline.autopilot.geo;

import ws.baseline.autopilot.GeoPoint;
import ws.baseline.autopilot.Paramotor;
import ws.baseline.autopilot.util.Convert;

import androidx.annotation.NonNull;
import java.util.Locale;

public class LandingZone {
    public final LatLngAlt destination;
    /** Landing direction in radians */
    public final double landingDirection;

    // Ground length of final approach
    public final double finalDistance = 150; // meters

    // Destination, as origin of coordinate system
    public final Point3V dest;

    public LandingZone(double lat, double lng, double alt, double landingDirection) {
        this.destination = new LatLngAlt(lat, lng, alt);
        this.landingDirection = landingDirection;
        this.dest = new Point3V(
                0,
                0,
                destination.alt,
                Math.sin(landingDirection),
                Math.cos(landingDirection),
                0
        );
    }

    /**
     * Landing pattern: start of final approach
     */
    public Point3V startOfFinal() {
        return new Point3V(
                -finalDistance * this.dest.vx,
                -finalDistance * this.dest.vy,
                dest.alt + finalDistance / Paramotor.glide,
                dest.vx,
                dest.vy,
                Paramotor.descentRate
        );
    }

    /**
     * Landing pattern: start of base leg
     */
    public Point3V startOfBase(int turn) {
        return new Point3V(
                -finalDistance * (this.dest.vx - turn * this.dest.vy),
                -finalDistance * (turn * this.dest.vx + this.dest.vy),
                dest.alt + 2 * finalDistance / Paramotor.glide,
                -dest.vx,
                -dest.vy,
                Paramotor.descentRate
        );
    }

    /**
     * Landing pattern: start of downwind leg
     */
    public Point3V startOfDownwind(int turn) {
        return new Point3V(
                finalDistance * turn * this.dest.vy,
                -finalDistance * turn * this.dest.vx,
                dest.alt + 3 * finalDistance / Paramotor.glide,
                -dest.vx,
                -dest.vy,
                Paramotor.descentRate
        );
    }

    /**
     * Convert lat, lng to x, y meters centered at current location
     */
    public Point3V toPoint3V(GeoPoint point) {
        final double bearing = Math.toRadians(Geo.bearing(this.destination.lat, this.destination.lng, point.lat, point.lng));
        final double distance = Geo.distance(this.destination.lat, this.destination.lng, point.lat, point.lng);
        return new Point3V(
                distance * Math.sin(bearing),
                distance * Math.cos(bearing),
                point.alt,
                point.vE,
                point.vN,
                point.climb
        );
    }

    /**
     * Convert x, y coordinates to lat, lng
     */
    public LatLng toLatLng(Point point) {
        final double bear = Math.toDegrees(Math.atan2(point.x, point.y));
        final double dist = Math.sqrt(point.x * point.x + point.y * point.y);
        return Geo.moveDirection(this.destination.lat, this.destination.lng, bear, dist);
    }

    @NonNull
    @Override
    public String toString() {
        return String.format(Locale.getDefault(), "%.6f, %.6f, %s, %.0f°", destination.lat, destination.lng, Convert.distance(destination.alt), Math.toDegrees(landingDirection));
    }
}

File: autopilot/lib/paralib/landingzone.cpp
#include <math.h>
#include <stdio.h>
#include "geo.h"
#include "landingzone.h"
#include "plan.h"

LandingZone::LandingZone(double lat, double lng, double alt, double landingDir) {
  destination = {
    .lat = lat,
    .lng = lng,
    .alt = alt
  };
  landingDirection = landingDir;
  dest = {
    .x = 0,
    .y = 0,
    .alt = 0,
    .vx = sin(landingDirection),
    .vy = cos(landingDirection),
    .climb = 0
  };
}

/**
 * Convert lat, lng to x, y meters centered at current location
 */
Point LandingZone::to_point(double lat, double lng) {
  const double bearing = geo_bearing(destination.lat, destination.lng, lat, lng);
  const double distance = geo_distance(destination.lat, destination.lng, lat, lng);
  Point point = {
    .x = distance * sin(bearing),
    .y = distance * cos(bearing)
  };
  return point;
}

/**
 * Convert lat, lng to x, y meters centered at current location
 */
Point3V LandingZone::to_point3V(GeoPointV *loc) {
  const double bearing = geo_bearing(destination.lat, destination.lng, loc->lat, loc->lng);
  const double distance = geo_distance(destination.lat, destination.lng, loc->lat, loc->lng);
  Point3V point = {
    .x = distance * sin(bearing),
    .y = distance * cos(bearing),
    .alt = loc->alt - destination.alt,
    .vx = loc->vE,
    .vy = loc->vN,
    .climb = loc->climb
  };
  return point;
}

/**
 * Landing pattern: start of final approach
 */
Point3V LandingZone::start_of_final() {
  Point3V point = {
    .x = -finalDistance * dest.vx,
    .y = -finalDistance * dest.vy,
    .alt = dest.alt + finalDistance / PARAMOTOR_GLIDE,
    .vx = dest.vx,
    .vy = dest.vy,
    .climb = PARAMOTOR_DESCENTRATE
  };
  return point;
}

/**
 * Landing pattern: start of base leg
 */
Point3V LandingZone::start_of_base(int turn) {
  Point3V point = {
    .x = -finalDistance * (dest.vx - turn * dest.vy),
    .y = -finalDistance * (turn * dest.vx + dest.vy),
    .alt = dest.alt + 2 * finalDistance / PARAMOTOR_GLIDE,
    .vx = -dest.vx,
    .vy = -dest.vy,
    .climb = PARAMOTOR_DESCENTRATE
  };
  return point;
}

/**
 * Landing pattern: start of base leg
 */
Point3V LandingZone::start_of_downwind(int turn) {
  Point3V point = {
    .x = finalDistance * turn * dest.vy,
    .y = -finalDistance * turn * dest.vx,
    .alt = dest.alt + 3 * finalDistance / PARAMOTOR_GLIDE,
    .vx = -dest.vx,
    .vy = -dest.vy,
    .climb = PARAMOTOR_DESCENTRATE
  };
  return point;
}
